pkg/graph/resolvers/system_intake_document.go

```diff
@@ -127,54 +127,59 @@ func CreateSystemIntakeDocument(
 //
 // Does *not* delete the uploaded file from S3, following the example of TRB request documents.
 func DeleteSystemIntakeDocument(ctx context.Context, store *storage.Store, id uuid.UUID) (*models.SystemIntakeDocument, error) {
-	if err := canDelete(ctx, store, id); err != nil {
+	if err := allowDelete(ctx, store, id); err != nil {
 		return nil, err
 	}
 
 	return store.DeleteSystemIntakeDocument(ctx, id)
 }
 
-// canView guards unauthorized views (downloads) of system intake documents
-// admins can view any document
-// requesters can view any document
-// GRB reviewers can view any document
-func canView(ctx context.Context, store *storage.Store, s3Key string) error {
+// CanViewDocument determines if a user can view a document
+func CanViewDocument(ctx context.Context, grbUsers []*models.SystemIntakeGRBReviewer, document *models.SystemIntakeDocument) bool {
 	// admins can view
 	if services.HasRole(ctx, models.RoleEasiGovteam) {
-		return nil
+		return true
 	}
 
+	// requester can view
+	user := appcontext.Principal(ctx).Account()
+	if document.GetCreatedBy() == user.Username {
+		return true
+	}
+
+	if isGRBViewer := slices.ContainsFunc(grbUsers, func(reviewer *models.SystemIntakeGRBReviewer) bool {
+		return reviewer.UserID == user.ID
+	}); isGRBViewer {
+		return true
+	}
+	return false
+}
+
+// canView guards unauthorized views (downloads) of system intake documents
+// admins can view any document
+// requesters can view any document
+// GRB reviewers can view any document
+func allowView(ctx context.Context, store *storage.Store, s3Key string) error {
 	document, err := store.GetSystemIntakeDocumentByS3Key(ctx, s3Key)
 	if err != nil {
-
 		if errors.Is(err, sql.ErrNoRows) {
 			// if the document was deleted, don't error and break
 			return nil
 		}
-
 		return err
 	}
 
-	// requester can view
-	user := appcontext.Principal(ctx).Account()
-	if document.GetCreatedBy() == user.Username {
-		return nil
-	}
```

Comment:
the reason i did this up here was to avoid a DB roundtrip if not needed

Reply:
Yes, I figured as much. I moved it as a compromise when splitting the functions between checking if the user is able to take an action and actually allowing the action.